### ISSUE TITLE
`is_sindex_sort` may be used uninitialized on GCC 4.8.3 (CentOS 7)

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#5073]
On `next`
```
In file included from ./src/rdb_protocol/datum.hpp:26:0,
                 from ./src/rdb_protocol/batching.hpp:9,
                 from ./src/rdb_protocol/shards.hpp:15,
                 from src/rdb_protocol/shards.cc:2:
./src/rdb_protocol/error.hpp: In member function 'virtual void ql::to_array_t::add_res(ql::env_t*, ql::result_t*, sorting_t)':
./src/rdb_protocol/error.hpp:208:19: warning: '*((void*)& is_sindex_sort +1)' may be used uninitialized in this function [-Wmaybe-uninitialized]
         if (!(test)) {                                         \
                   ^
src/rdb_protocol/shards.cc:401:39: note: '*((void*)& is_sindex_sort +1)' was declared here
                 boost::optional<bool> is_sindex_sort;
```

**Comment by mlucy** [comment]
I can see why the compiler is confused here.  This code is only correct because we know that if we have no substreams then `v` is empty.  We have a sanity check that `is_sindex_sort` is actually filled in, but I guess the compiler isn't smart enough to realize that we'll abort if that check fails?

**Comment by danielmewes** [comment]
My impression was that the compiler actually complained about the `r_sanity_check(is_sindex_sort)`. But I could be wrong (it doesn't seem to provide a line number in shards.cc).

**Comment by mlucy** [comment]
Oh, in that case it's just weird then, since it isn't complaining about the `if (is_sindex_sort)` a few lines above.

**Comment by danielmewes** [comment]
This also fails on a bunch of Ubuntu build targets. Looking into it now.

**Comment by danielmewes** [comment]
Fixed by 18ef496
